### ISSUE TITLE
Shuffle algorithm reapeating songs frecuently

### ISSUE THREAD

**Issue by apika** [vanilla-music/vanilla#475]
Is possible that the shuffle option is not really that shuffle? After some time listening in this mode or when reseting the phone, I found that some songs shuffle "together".
Does Vanilla algorithm works like Winamp's shuffle? When You shuffle in Winamp (at least old versions), It will jump X number ahead in the list so using the same list and starting from the same point will give you the same songs order, 

Sorry for not giving a lot of feedback, I found this very difficult to debug.

**Comment by adrian-bl** [comment]
Just to make sure: Are you talking about shuffling the timeline (eg: the button on the left side) or the Random mode (the dieces on the right side) ?

The random mode creates an internal playlist with all songs and will pop() one of them if you reached the last item on the playlist.

**Comment by apika** [comment]
I am talking about the Random mode (because some programs translate Shuffle as Random but in spanish). The left button doesn't affect this situation.

**Comment by adrian-bl** [comment]
> I am talking about the Random mode,

Ok, thanks for clearing this up.

> It will jump X number ahead in the list so using the same list and starting from the same point will 
> give you the same songs order,

No: Our Random mode works different:

Vanilla Music fetches all songs from the library and keeps them in a shuffled array. Every time the end of the queue is reached, the last item of the shuffled array is removed and appended to the queue.

So songs should will only start to repeat after all songs in the library have been played.

**Comment by apika** [comment]
> So songs should will only start to repeat after all songs in the library have been played.

Just today I hear Anyway from Genesis twice in less than 2 hours of non continuos music playing. I have a lot of music in my SD, for sure more than 2 hours (cause prog rock, 6 tracks are 2 hours for example). I didn't reset my phone in the middle.

Does this shuffled array resets every time I stop playing music? Theres a way to mantain this array even when closing the app/stop playing music?.

**Comment by adrian-bl** [comment]
> Does this shuffled array resets every time I stop playing music? Theres a way to mantain this array 
> even when closing the app/stop playing music?.

No, but it will reset if Vanilla Music gets killed while being in the background, which is rare but might happen (depending on your OS and Hardware)

**Comment by apika** [comment]
There is a way to know that? Something like the queue not remembering your last song? Or the notification closing even when you left it to be always on?

**Comment by adrian-bl** [comment]
> There is a way to know that? Something like the queue not remembering your last song?

Hmm, no: such restarts are somewhat hard to spot: One way to tell is if the scrolling position in the library view jumped back to the top

**Comment by hboetes** [comment]
getting random songs is not as easy as it sounds: http://www.independent.co.uk/life-style/gadgets-and-tech/news/why-random-shuffle-feels-far-from-random-10066621.html

You basically have to take notes for how often you have played songs and try to play each song as often as the other. And take changes in the collection into account so you have to store the first time the track showed up in the collection as well.

I would welcome that feature.

**Comment by hboetes** [comment]
Awesome that you are working on implementing this!

**Comment by adrian-bl** [comment]
You misunderstood me: i am not working on this, i just said that the required data is already there

**Comment by hboetes** [comment]
I’m trying to inspire you. :-)

**Comment by Adonai** [comment]
Lol, this is actually very funny issue :D

Go ahead, tell me that I miss entropy in my `/proc/sys/kernel/random/entropy_avail` so I should prematurely click at random places of screen every time I want my song to be truly random XD

**Comment by hboetes** [comment]
ok, let me explain the issue for you, look at this code:

  `for i in {1..100}; do echo $(($RANDOM % 100 + 1)); done|sort -n|uniq -c|sort -n`

Do you really want some of your 100 tracks to be played 3, 4 times and others not at all? True randomness doesn’t help with that issue. Most people want their tracks to be played equally often but in random order.

**Comment by Adonai** [comment]
Ok, that makes sense.

But we maintain a list of available songs in random order and just walk through it in this mode, so songs aren't repeated.

The only issue I can see is that when this list ends and next song is requested, the list is reshuffled and may provide the same song as the one that was last (or near enough to the end of old list) as first one. This may give impression of imperfect random mode.

**Comment by hboetes** [comment]
Oh that’s interesting. Where is that list stored?

**Comment by Adonai** [comment]
IIRC it's somewhere in MediaUtils.getRandomSong

**Comment by hboetes** [comment]
No results found for MediaUtils.getRandomSong

**Comment by mindforger** [comment]
you could double split your list in 3 pieces

1st, 2nd and 3rd part (which i will call buffer)

when finished with the 1st, you scramble a new 1st from half the 1st and the buffer, left over songs get dropped in buffer

when finished with the 2nd, you scrambe a new 2nd with half the songs from buffer and the 2nd part, leftovers back to buffer

the distance between a song repeating is never smaller than a third of the list

**Comment by adrian-bl** [comment]
The current implementation (see getRandomSong) grabs a copy of the library, shuffles it and pops a song on every call.

We are not repeating songs until the random list is cleared (app restart) or exhausted

**Comment by hboetes** [comment]
Right, that explains my issue. As soon as I stop the car the app is restarted. Would it be possible to store the library so it survives restarts?

**Comment by apika** [comment]
You could make you list (if you want all the songs just click in play all), then shuffle tracks and save that list. That's how I fix my "problem".

**Comment by SebiderSushi** [comment]
> Vanilla Music fetches all songs from the library and keeps them in a shuffled array. Every time the end of the queue is reached, the last item of the shuffled array is removed and appended to the queue.
> 
> So songs should will only start to repeat after all songs in the library have been played.

Wait - i always thought it would be the other way around?
Because this behavior sounds a lot like Shuffle & Repeat, which also reshuffles the list once all songs have been played. The only differences being that you can have can always see all passed and future tracks in the queue instead of 20 passed and the unplayed tracks and that it does not get reset on app restarts, which, as i read it, was mostly experienced as a (possibly unwanted) side effect rather than a feature.

With a shuffle mode like this i always presumed that "Random" would in contrast mean that every track from the library would have a statistically equal chance of being the next one.

What sets the Random mode apart from Shuffle & Repeat in regards to plaing dynamics?
